torch/csrc/jit/codegen/cuda/contiguity.cpp

```diff
@@ -32,150 +468,112 @@ ContigIDs::ContigIDs(
       " != ",
       root_contiguity_.size());
 
-  // GpuLower is required to honor halo constraints
-  if (!ignore_halo_constraint) {
-    TORCH_INTERNAL_ASSERT(GpuLower::hasCurrent(), "GpuLower not found");
-  }
-
-  for (const auto i : c10::irange(root_domain_.size())) {
-    auto root_domain_i = root_domain_[i]->as<IterDomain>();
-    root_to_indexed_id_[root_domain_i] = root_domain_i;
+  for (const auto root_domain_i : c10::irange(root_domain_.size())) {
+    auto root_domain_id = root_domain_[root_domain_i]->as<IterDomain>();
+    root_to_indexed_id_[root_domain_id] = root_domain_id;
     // Initialize to false
-    is_contig_root_[root_domain_i] = false;
+    is_contig_root_[root_domain_id] = false;
     // If a root domain has halo, can't use merged domain even if
     // both inputs are contiguous. HaloInfo is also initialized for
     // rfactor root domains, which should just return "zero"
     // RootAxisInfo. This should be safe as no rfactor tensor should
     // need halo.
-    if (root_contiguity_[i] &&
-        (ignore_halo_constraint ||
-         !GpuLower::current()
-              ->haloInfo()
-              ->getRootAxisInfo(root_domain_i)
-              .hasHalo())) {
-      contig_ids_.emplace(root_domain_i);
-      is_contig_root_[root_domain_i] = true;
-      within_contig_ids_[root_domain_i] = std::unordered_set<IterDomain*>();
+    if (root_contiguity_[root_domain_i] &&
+        !halo_info_->getRootAxisInfo(root_domain_id).hasHalo()) {
+      contig_ids_.emplace(root_domain_id);
+      is_contig_root_[root_domain_id] = true;
+      within_contig_ids_[root_domain_id] = std::unordered_set<IterDomain*>();
     }
   }
 
   if (!contig_ids_.empty()) {
-    auto exprs = StmtSort::getExprs(ids[0]->fusion(), {ids.begin(), ids.end()});
+    auto exprs = StmtSort::getExprs(
+        ids[0]->fusion(),
+        {ids.begin(), ids.end()},
+        {root_domain_.begin(), root_domain_.end()});
     for (auto expr : exprs) {
       handle(expr);
     }
   }
 }
 
 void ContigIDs::handle(Merge* merge) {
-  // If either input is non-contiguous so is output.
-  const auto inner = merge->inner();
-  const auto outer = merge->outer();
-  const auto out = merge->out();
-
-  if (!isContig(inner) || !isContig(outer)) {
+  // If output is not consistently ordered or doesn't solely consume all root
+  // domains in its dependencies, then it can't be a contiguously indexable
+  // iterdomain.
+  if (!(consistent_transform_info_->isConsistentlyOrdered(merge->out()) &&
+        consistent_transform_info_->exclusivelyConsumesRoots(merge->out()))) {
```

Comment:
Is the exclusiveness always necessary? For example:

`[4, 8] -> split(0, 2) -> [4/2, 2, 8] -> merge(1) ->[4/2, 2*8]`

In this case the second leaf ID (`2*8`) doesn't exclusively consume the first root ID, but it should be valid to do indexing at that domain. 

However, to do so, we would also need to index the first leaf ID, which is the outer output domain of the split, so some more extension would be needed.

Reply:
I suspect the closer to the leaf domains, the more efficient the index math would be. 

Would be good to have this as a comment on the code.

Reply:
@Sergei-Lebedev I know it's early in your understanding of indexing, but this is a really interesting point we should review eventually.